rfc/20241118-module-vars-and-outputs-deprecation.md

```diff
@@ -0,0 +1,178 @@
+should accept a required string value for a warning message:
+
+```hcl
+variable "this_is_my_variable" {
+  type        = string
+  description = "This is a variable for the old way of configuring things."
+  deprecated  = "This variable will be removed on 2024-12-31. Use that_is_my_variable instead."
+}
+
+output "this_is_my_output" {
+  value       = some_resource.some_name.some_field
+  description = "This is an output for the old way of using things."
+  deprecated  = "This output will be removed on 2024-12-31. Use that_is_my_output instead."
+}
+```
+
+#### Deprecated module variable warning
+
+If module caller specify non-null value for a deprecated variable, OpenTofu should produce a deprecation warning as 
+specified by module author:
+
+```hcl
+│ Warning: The variable "this_is_my_variable" is marked as deprecated by module author.
+│ 
+│   on mod/main.tf line 9, in module call "mod":
+│    9:     this_is_my_variable = "something"
+│ 
+│ This variable will be removed on 2024-12-31. Use that_is_my_variable instead.
+```
+
+> [!NOTE]
+> It is possible for module user to specify a null value for a deprecated variable and receive no warnings. OpenTofu treats
+> variables with explicit null values the same way as if it was ommited completely. This is not an ideal, however an idiomatic
+> OpenTofu language approach. See `null` values note [in the docs](https://opentofu.org/docs/language/expressions/types/#types).
+
+#### Deprecated module output warning
+
+If module caller has references to deprecated module outputs in their configuration, OpenTofu should produce a
+deprecation warning as specified by module author:
+
+```hcl
+│ Warning: The output "this_is_my_output" is marked as deprecated by module author.
+│ 
+│   on main.tf line 9, in resource "some_resource" "some_name":
+│    9:     some_field = module.mod.this_is_my_output
+│ 
+│ This output will be removed on 2024-12-31. Use that_is_my_output instead.
+```
+
+> [!NOTE]
+> Module users may rely on [output precondition check](https://opentofu.org/docs/language/values/outputs/#custom-condition-checks).
+> However it doesn't require implicit references, so this type of usage will not be counted when generating deprecation warnings. We
+> might want to forbid deprecation of outputs with precondition checks so module authors would be required to migrate them beforehand.
+
+### Technical Approach
+
+Technical approach should be different for input variables and outputs. For input variables, we can follow existing approach, where
+variables are being validated.
+
+As for module outputs, it is slightly harder to determine if it's is actually referenced in the user configuration, because
+`nodeExpandOutput` is included in the graph regardless of its actual usage. On graph walk, this node expands to one or more
+`NodeApplyableOutput` nodes. `NodeApplyableOutput` also run validation of precondition checks on its execution.
```

Comment:
Handling this at the graph building stage will have all of the limitations I mentioned in my previous comment, and so I expect it wouldn't work reliably enough for module authors to trust that most/all of their users would've seen the warning. :confounded: 

A different approach we could consider is to treat this a little like how we handle "sensitive values", which is a dynamic-analysis solution based on [`cty`'s Value Marks](https://github.com/zclconf/go-cty/blob/main/docs/marks.md).

Specifically, we could decide that the value produced by any input variable or output value that has a `deprecated` argument is "marked" with a new marking representing that the value was derived from a deprecated source, and then introduce checks at various places in the codebase where values are finally used (e.g. in the arguments of `resource`/`data`/`module`/`provider`, blocks, or the `value` of an `output` block) that use [`Value.UnmarkDeep`](https://pkg.go.dev/github.com/zclconf/go-cty@v1.15.1/cty#Value.UnmarkDeep) to determine if any value anywhere in the resulting data structure is marked as having come from a deprecated source.

Whereas the existing `marks.Sensitive` is effectively just a boolean flag of either sensitive (marked) or not (unmarked), I think `marks.Deprecated` would need to be a type that carries with it the address of the thing that was deprecated, so that the address will follow the deprecated values around through intermediate expressions and then the final error message can mention specifically what was deprecated:

```
│ Warning: Value derived from a deprecated source
│ 
│   on main.tf line 9, in resource "some_resource" "some_name":
│    9:     some_field = local.this_is_my_output
│ 
│ This value is derived from module.mod.output.this_is_my_output, which is
│ deprecated with the following message:
│ 
│ This output will be removed on 2024-12-31. Use that_is_my_output instead.
```

Although this dynamic approach is admittedly slightly more messy because we'd need to add the checks at the place where the values are finally _used_, rather than the places they originally came from, I think this does have the advantage of allowing us to use the same strategy for deprecations of all kinds. If the provider protocol were to expand to allow provider authors to mark attributes as deprecated in future then that too could be represented as a marked value and then caught and reported by the same downstream checks.

Doing this via dynamic analysis also makes the detection more precise, although I'm not sure if this is actually an advantage or a disadvantage: `false ? module.mod.this_is_my_output : "fallback"` would discard the marked value and therefore cause it to not be reported. This seems like a similar situation to how explicitly assigning `null` to a deprecated input variable would not generated a warning: it effectively hides a reference that could still cause an error later. :thinking: